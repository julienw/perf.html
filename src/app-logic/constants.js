/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

// The current version of the Gecko profile format.
export const GECKO_PROFILE_VERSION = 16;

// The current version of the "processed" profile format.
export const PROCESSED_PROFILE_VERSION = 23;

// The following are the margin sizes for the left and right of the timeline. Independent
// components need to share these values.
export const TIMELINE_MARGIN_RIGHT = 15;
export const TIMELINE_MARGIN_LEFT = 150;

export const TIMELINE_SETTINGS_HEIGHT = 26;

// Export the value for tests, and for computing the max height of the timeline
// for the splitter.
export const TRACK_SCREENSHOT_HEIGHT = 50;

// The following values are for network track.
export const TRACK_NETWORK_ROW_HEIGHT = 5;
export const TRACK_NETWORK_ROW_REPEAT = 7;
export const TRACK_NETWORK_HEIGHT =
  TRACK_NETWORK_ROW_HEIGHT * TRACK_NETWORK_ROW_REPEAT;

// The following values are for memory track.
export const TRACK_MEMORY_GRAPH_HEIGHT = 25;
export const TRACK_MEMORY_MARKERS_HEIGHT = 15;
export const TRACK_MEMORY_HEIGHT =
  TRACK_MEMORY_GRAPH_HEIGHT + TRACK_MEMORY_MARKERS_HEIGHT;
export const TRACK_MEMORY_LINE_WIDTH = 2;

// Height of the blank area in process track.
export const TRACK_PROCESS_BLANK_HEIGHT = 30;

// Height of timeline ruler.
export const TIMELINE_RULER_HEIGHT = 20;
