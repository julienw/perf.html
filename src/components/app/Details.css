.Details {
  flex: auto;

  display: flex;
  flex-direction: column;
}
