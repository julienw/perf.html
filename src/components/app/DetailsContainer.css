.DetailsContainer .layout-pane > * {
  box-sizing: border-box;
  width: 100%;
  height: 100%;
}

.DetailsContainer .layout-pane:not(.layout-pane-primary) {
  max-width: 250px;
}

/* overriding defaults from splitter-layout */
.DetailsContainer {
  /* SplitterLayout injects position: absolute, this conflicts with our using
   * Flex Layout. */
  position: unset;
}

.DetailsContainer .layout-splitter {
  background: var(--grey-10); /* Same background as sidebars */
  border-left: 1px solid var(--grey-40);
}

.DetailsContainer .layout-splitter:hover {
  background: var(--grey-40); /* same as the border above */
}


