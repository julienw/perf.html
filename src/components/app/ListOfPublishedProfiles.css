/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.publishedProfilesList {
  padding: 0;
  margin: 0;
  list-style: none;
}

.publishedProfilesListItem {
  padding: 0;
}

.publishedProfilesLink {
  display: flex;
  padding: 4px 8px; /* 4px as vertical padding so that it's 8 when added up, and 8px for the left and right padddings */
  color: inherit;
  text-decoration: none;
}

.publishedProfilesLink:hover {
  background: var(--grey-90-a05);
}

.publishedProfilesDate {
  width: 10em; /* The default flex properties make that it can shrink, so this is really the initial size. */
  padding-right: 4px;
  white-space: nowrap;
}

.publishedProfilesInfo {
  min-width: 0;
}

.publishedProfilesName {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}
