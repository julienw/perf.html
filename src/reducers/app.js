/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';

import { getSelectedTab } from './url-state';
import { tabSlugs } from '../app-logic/tabs-handling';

import type { TabSlug } from '../app-logic/tabs-handling';
import type { Action } from '../types/store';
import type {
  State,
  AppState,
  AppViewState,
  IsSidebarOpenPerPanelState,
  Reducer,
} from '../types/reducers';

function view(
  state: AppViewState = { phase: 'INITIALIZING' },
  action: Action
): AppViewState {
  if (state.phase === 'DATA_LOADED') {
    // Let's not come back at another phase if we're already displaying a profile
    return state;
  }

  switch (action.type) {
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_STORE':
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_URL':
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_ADDON':
      return {
        phase: 'INITIALIZING',
        additionalData: {
          message: action.error.message,
          attempt: action.error.attempt,
        },
      };
    case 'ERROR_RECEIVING_PROFILE_FROM_FILE':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_ADDON':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_STORE':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_URL':
      return { phase: 'FATAL_ERROR', error: action.error };
    case 'WAITING_FOR_PROFILE_FROM_ADDON':
      return { phase: 'INITIALIZING' };
    case 'ROUTE_NOT_FOUND':
      return { phase: 'ROUTE_NOT_FOUND' };
    case 'RECEIVE_ZIP_FILE':
    case 'VIEW_PROFILE':
      return { phase: 'DATA_LOADED' };
    default:
      return state;
  }
}

function isUrlSetupDone(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'URL_SETUP_DONE':
      return true;
    default:
      return state;
  }
}

function hasZoomedViaMousewheel(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'HAS_ZOOMED_VIA_MOUSEWHEEL': {
      return true;
    }
    default:
      return state;
  }
}

function isSidebarOpenPerPanel(
  state: IsSidebarOpenPerPanelState,
  action: Action
): IsSidebarOpenPerPanelState {
  if (state === undefined) {
    state = {};
    tabSlugs.forEach(tabSlug => (state[tabSlug] = false));
  }

  switch (action.type) {
    case 'CHANGE_SIDEBAR_OPEN_STATE': {
      const { tab, isOpen } = action;
      // Due to how this action will be dispatched we'll always have the value
      // changed so we don't need the performance optimization of checking the
      // stored value against the new value.
      return {
        ...state,
        [tab]: isOpen,
      };
    }
    default:
      return state;
  }
}

/**
 * The panels that make up the timeline, details view, and sidebar can all change
 * their sizes depending on the state that is fed to them. In order to control
 * the invalidations of this sizing information, provide a "generation" value that
 * increases monotonically for any change that potentially changes the sizing of
 * any of the panels. This provides a mechanism for subscribing components to
 * deterministically update their sizing correctly.
 */
function panelLayoutGeneration(state: number = 0, action: Action): number {
  switch (action.type) {
    case 'INCREMENT_PANEL_LAYOUT_GENERATION':
    // Sidebar: (fallthrough)
    case 'CHANGE_SIDEBAR_OPEN_STATE':
    // Timeline: (fallthrough)
    case 'HIDE_GLOBAL_TRACK':
    case 'SHOW_GLOBAL_TRACK':
    case 'ISOLATE_PROCESS':
    case 'ISOLATE_PROCESS_MAIN_THREAD':
    case 'HIDE_LOCAL_TRACK':
    case 'SHOW_LOCAL_TRACK':
    case 'ISOLATE_LOCAL_TRACK':
    // Committed range changes: (fallthrough)
    case 'COMMIT_RANGE':
    case 'POP_COMMITTED_RANGES':
      return state + 1;
    default:
      return state;
  }
}

/**
 * Clicking on tracks can switch between different tabs. This piece of state holds
 * on to the last relevant thread-based tab that was viewed. This makes the UX nicer
 * for when a user clicks on a Network track, and gets taken to the Network
 * panel, then clicks on a thread or process track. With this state we can smoothly
 * transition them back to the panel they were using.
 */
function lastVisibleThreadTabSlug(
  state: TabSlug | null = null,
  action: Action
): TabSlug | null {
  switch (action.type) {
    case 'SELECT_TRACK':
    case 'CHANGE_SELECTED_TAB':
      if (action.selectedTab !== 'network-chart') {
        return action.selectedTab;
      }
      return state;
    default:
      return state;
  }
}

const appStateReducer: Reducer<AppState> = combineReducers({
  view,
  isUrlSetupDone,
  hasZoomedViaMousewheel,
  isSidebarOpenPerPanel,
  panelLayoutGeneration,
  lastVisibleThreadTabSlug,
});

export default appStateReducer;

export const getApp = (state: State): AppState => state.app;
export const getView = (state: State): AppViewState => getApp(state).view;
export const getIsUrlSetupDone = (state: State): boolean =>
  getApp(state).isUrlSetupDone;
export const getHasZoomedViaMousewheel = (state: State): boolean => {
  return getApp(state).hasZoomedViaMousewheel;
};
export const getIsSidebarOpen = (state: State): boolean =>
  getApp(state).isSidebarOpenPerPanel[getSelectedTab(state)];
export const getPanelLayoutGeneration = (state: State) =>
  getApp(state).panelLayoutGeneration;
export const getLastVisibleThreadTabSlug = (state: State) =>
  getApp(state).lastVisibleThreadTabSlug;
