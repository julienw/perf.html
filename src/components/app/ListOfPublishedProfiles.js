/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import memoize from 'memoize-immutable';
import classNames from 'classnames';

import { InnerNavigationLink } from 'firefox-profiler/components/shared/InnerNavigationLink';
import { ProfileMetaInfoSummary } from 'firefox-profiler/components/shared/ProfileMetaInfoSummary';
import {
  ButtonWithPanel,
  ConfirmDialog,
} from 'firefox-profiler/components/shared/ButtonWithPanel';

import {
  listAllProfileData,
  deleteProfileData,
  type ProfileData,
} from 'firefox-profiler/app-logic/published-profiles-store';
import { formatSeconds } from 'firefox-profiler/utils/format-numbers';
import { deleteProfileOnServer } from 'firefox-profiler/profile-logic/profile-store';

import type { Milliseconds, StartEndRange } from 'firefox-profiler/types/units';

import './ListOfPublishedProfiles.css';

// This component displays all published profile, and makes it possible to load
// them by clicking on them, or delete them.

const ONE_DAY_IN_MS = 24 * 60 * 60 * 1000;
const ONE_YEAR_IN_MS = 365 * ONE_DAY_IN_MS;

// All formats are lazy-initialized at the first use by using the `memoize`
// tool, because the DateTimeFormat constructor is slow.
// Because there's no parameter to these functions there will only ever be
// memoized once and should be fairly efficient. We can replace with a dedicated
// "lazy" function in the future if performance ever becomes a problem.
const dateFormats = {
  thisDay: memoize(
    () =>
      new Intl.DateTimeFormat(undefined, { hour: 'numeric', minute: 'numeric' })
  ),
  thisYear: memoize(
    () =>
      new Intl.DateTimeFormat(undefined, {
        month: 'short',
        day: 'numeric',
        hour: 'numeric',
        minute: 'numeric',
      })
  ),
  ancient: memoize(
    () =>
      new Intl.DateTimeFormat(undefined, {
        year: 'numeric',
        month: 'short',
        day: 'numeric',
      })
  ),
};

function _formatDate(date: Date, nowTimestamp: Milliseconds) {
  const timeDifference = nowTimestamp - +date;
  if (timeDifference < 0 || timeDifference > ONE_YEAR_IN_MS) {
    return dateFormats.ancient().format(date);
  }
  if (timeDifference > ONE_DAY_IN_MS) {
    return dateFormats.thisYear().format(date);
  }
  return dateFormats.thisDay().format(date);
}

function _formatRange(range: StartEndRange): string {
  return formatSeconds(range.end - range.start, 3, 1);
}

type PublishedProfileProps = {|
  +profileData: ProfileData,
  +nowTimestamp: Milliseconds,
  +withActionButtons: boolean,
|};

type PublishedProfileState = {|
  +confirmDialogIsOpen: boolean,
  +hasBeenDeleted: boolean,
|};

class PublishedProfile extends React.PureComponent<
  PublishedProfileProps,
  PublishedProfileState
> {
  state = { confirmDialogIsOpen: false, hasBeenDeleted: false };
  _deleteButtonRef = React.createRef();
  _actionButtonsRef = React.createRef();

  onConfirmDeletion = async () => {
    const { profileToken, jwtToken } = this.props.profileData;
    if (!jwtToken) {
      throw new Error(
        `We have no JWT token for this profile, so we can't delete it. This shouldn't happen.`
      );
    }
    await deleteProfileOnServer({ profileToken, jwtToken });
    await deleteProfileData(profileToken);
    this.setState({ hasBeenDeleted: true });
  };

  onCloseConfirmDialog = () => {
    this.setState({ confirmDialogIsOpen: false });

    // Let's focus the delete button after dismissing the dialog, but _only_ if
    // the focus was part of the dialog before.
    // Note: we might need to get a more precise ref to the right
    // buttonWithPanel wrapper when we'll have more buttons.
    if (this._deleteButtonRef.current && this._actionButtonsRef.current) {
      const deleteButton = this._deleteButtonRef.current;
      const actionButtons = this._actionButtonsRef.current;
      if (actionButtons.matches(':focus-within')) {
        deleteButton.focus();
      }
    }
  };

  onOpenConfirmDialog = () => {
    this.setState({ confirmDialogIsOpen: true });
  };

  render() {
    const { profileData, nowTimestamp, withActionButtons } = this.props;
    const { hasBeenDeleted, confirmDialogIsOpen } = this.state;

    if (hasBeenDeleted) {
      return null;
    }

    let { urlPath } = profileData;
    if (!urlPath.startsWith('/')) {
      urlPath = '/' + urlPath;
    }
    const location = `${window.location.origin}/${urlPath}`;
    const slicedProfileToken = profileData.profileToken.slice(0, 6);
    const profileName = profileData.name
      ? profileData.name
      : `Profile #${slicedProfileToken}`;
    const smallProfileName = profileData.name
      ? profileData.name
      : '#' + slicedProfileToken;

    return (
      <li
        className={classNames('publishedProfilesListItem', {
          publishedProfilesListItem_ConfirmDialogIsOpen: confirmDialogIsOpen,
        })}
      >
        <a
          className="publishedProfilesLink"
          href={location}
          title={`Click here to load profile ${smallProfileName}`}
        >
          <div className="publishedProfilesDate">
            {_formatDate(profileData.publishedDate, nowTimestamp)}
          </div>
          <div className="publishedProfilesInfo">
            <div className="publishedProfilesName">
              <strong>{profileName}</strong> (
              {_formatRange(profileData.publishedRange)})
            </div>
            <ProfileMetaInfoSummary meta={profileData.meta} />
          </div>
        </a>
        {withActionButtons ? (
          <div
            className="publishedProfilesActionButtons"
            ref={this._actionButtonsRef}
          >
            {profileData.jwtToken ? (
              <ButtonWithPanel
                buttonRef={this._deleteButtonRef}
                buttonClassName="publishedProfilesDeleteButton photon-button photon-button-default"
                label="Delete"
                title={`Click here to delete the profile ${smallProfileName}`}
                onPanelOpen={this.onOpenConfirmDialog}
                onPanelClose={this.onCloseConfirmDialog}
                panelContent={
                  <ConfirmDialog
                    title={`Delete ${profileName}`}
                    confirmButtonText="Delete"
                    confirmButtonType="destructive"
                    cancelButtonText="Cancel"
                    onConfirmButtonClick={this.onConfirmDeletion}
                  >
                    Are you sure you want to delete uploaded data for this
                    profile? Links for shared copies will no longer work.
                  </ConfirmDialog>
                }
              />
            ) : (
              <button
                className="publishedProfilesDeleteButton photon-button photon-button-default"
                type="button"
                title="This profile cannot be deleted because we lack the authorization information."
                disabled
              >
                Delete
              </button>
            )}
          </div>
        ) : null}
      </li>
    );
  }
}

type Props = {|
  withActionButtons: boolean,
  limit?: number,
|};

type State = {|
  profileDataList: null | ProfileData[],
|};

export class ListOfPublishedProfiles extends PureComponent<Props, State> {
  state = {
    profileDataList: null,
  };

  async componentDidMount() {
    const profileDataList = await listAllProfileData();

    // It isn't ideal to use a setState here, but this is the only way.
    // eslint-disable-next-line react/no-did-mount-set-state
    this.setState({
      // We want to display the list with the most recent uploaded profile first.
      profileDataList: profileDataList.reverse(),
    });
  }

  render() {
    const { limit, withActionButtons } = this.props;
    const { profileDataList } = this.state;

    if (!profileDataList) {
      return null;
    }

    if (!profileDataList.length) {
      return (
        <p className="photon-body-30">No profile has been published yet!</p>
      );
    }

    const reducedProfileDataList = limit
      ? profileDataList.slice(0, limit)
      : profileDataList;

    const profilesRestCount =
      profileDataList.length - reducedProfileDataList.length;

    const nowTimestamp = Date.now();

    return (
      <>
        <ul className="publishedProfilesList">
          {reducedProfileDataList.map(profileData => (
            <PublishedProfile
              key={profileData.profileToken}
              profileData={profileData}
              nowTimestamp={nowTimestamp}
              withActionButtons={withActionButtons}
            />
          ))}
        </ul>
        {profilesRestCount > 0 ? (
          <p>
            <InnerNavigationLink dataSource="uploaded-recordings">
              See and manage all your recordings ({profilesRestCount} more)
            </InnerNavigationLink>
          </p>
        ) : null}
      </>
    );
  }
}
