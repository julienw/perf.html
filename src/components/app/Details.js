/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import explicitConnect from '../../utils/connect';
import TabBar from './TabBar';
import ProfileCallTreeView from '../calltree/ProfileCallTreeView';
import MarkerTable from '../marker-table';
import StackChart from '../stack-chart/';
import MarkerChart from '../marker-chart/';
import FlameGraph from '../flame-graph/';
import { changeSelectedTab, changeTabOrder } from '../../actions/app';
import { getTabOrder } from '../../reducers/profile-view';
import { getSelectedTab } from '../../reducers/url-state';
import CallNodeContextMenu from '../shared/CallNodeContextMenu';
import MarkerTableContextMenu from '../marker-table/ContextMenu';
import ProfileThreadHeaderContextMenu from '../header/ProfileThreadHeaderContextMenu';
import { toValidTabSlug } from '../../utils/flow';

import type { Tab } from './TabBar';
import type {
  ExplicitConnectOptions,
  ConnectedProps,
} from '../../utils/connect';
import type { TabSlug } from '../../types/actions';

require('./Details.css');

type StateProps = {|
  +tabOrder: number[],
  +selectedTab: TabSlug,
|};

type DispatchProps = {|
  +changeSelectedTab: typeof changeSelectedTab,
  +changeTabOrder: typeof changeTabOrder,
|};

type Props = ConnectedProps<{||}, StateProps, DispatchProps>;

class ProfileViewer extends PureComponent<Props> {
  // If updating this list, make sure and update the tabOrder reducer with another index.
  _tabs: Tab[] = [
    {
      name: 'calltree',
      title: 'Call Tree',
    },
    {
      name: 'flame-graph',
      title: 'Flame Graph',
    },
    {
      name: 'stack-chart',
      title: 'Stack Chart',
    },
    {
      name: 'marker-chart',
      title: 'Marker Chart',
    },
    {
      name: 'marker-table',
      title: 'Marker Table',
    },
    {
      name: 'network-chart',
      title: 'Network',
    },
  ];

  _onSelectTab = (selectedTab: string) => {
    const { changeSelectedTab } = this.props;
    const tabSlug = toValidTabSlug(selectedTab);
    if (!tabSlug) {
      throw new Error('Attempted to change to a tab that does not exist.');
    }
    changeSelectedTab(tabSlug);
  };

  render() {
    const { tabOrder, changeTabOrder, selectedTab } = this.props;
    return (
      <div className="Details">
        <TabBar
          tabs={this._tabs}
          selectedTabName={selectedTab}
          tabOrder={tabOrder}
          onSelectTab={this._onSelectTab}
          onChangeTabOrder={changeTabOrder}
        />
        {
          {
            calltree: <ProfileCallTreeView />,
            'marker-table': <MarkerTable />,
            'stack-chart': <StackChart />,
            'marker-chart': <MarkerChart />,
            'network-chart': <MarkerChart />,
            'flame-graph': <FlameGraph />,
          }[selectedTab]
        }
        <CallNodeContextMenu />
        <MarkerTableContextMenu />
        <ProfileThreadHeaderContextMenu />
      </div>
    );
  }
}

const options: ExplicitConnectOptions<{||}, StateProps, DispatchProps> = {
  mapStateToProps: state => ({
    tabOrder: getTabOrder(state),
    selectedTab: getSelectedTab(state),
  }),
  mapDispatchToProps: {
    changeSelectedTab,
    changeTabOrder,
  },
  component: ProfileViewer,
};

export default explicitConnect(options);
