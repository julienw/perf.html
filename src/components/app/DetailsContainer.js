/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import React from 'react';
import SplitterLayout from 'react-splitter-layout';

import Details from './Details';
import selectSidebar from '../sidebar';

import { getSelectedTab } from '../../reducers/url-state';
import explicitConnect from '../../utils/connect';

import type { TabSlug } from '../../types/actions';
import type { ExplicitConnectOptions } from '../../utils/connect';

import './DetailsContainer.css';

function dispatchResizeEvent() {
  const event = new UIEvent('resize', { view: window });
  window.dispatchEvent(event);
}

type StateProps = {|
  +selectedTab: TabSlug,
|};

function DetailsContainer({ selectedTab }: StateProps) {
  const Sidebar = selectSidebar(selectedTab);

  return (
    <SplitterLayout
      customClassName="DetailsContainer"
      percentage
      secondaryInitialSize={20}
      onDragEnd={dispatchResizeEvent}
    >
      <Details />
      {Sidebar && <Sidebar />}
    </SplitterLayout>
  );
}

const options: ExplicitConnectOptions<{||}, StateProps, {||}> = {
  mapStateToProps: state => ({
    selectedTab: getSelectedTab(state),
  }),
  component: DetailsContainer,
};

export default explicitConnect(options);
