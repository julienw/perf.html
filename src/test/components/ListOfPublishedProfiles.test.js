/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import { ListOfPublishedProfiles } from 'firefox-profiler/components/app/ListOfPublishedProfiles';
import { storeProfileData } from 'firefox-profiler/app-logic/published-profiles-store';
import { blankStore } from '../fixtures/stores';
import { mockDate } from '../fixtures/mocks/date';

import 'fake-indexeddb/auto';
import FDBFactory from 'fake-indexeddb/lib/FDBFactory';

function resetIndexedDb() {
  // This is the recommended way to reset the IDB state between test runs, but
  // neither flow nor eslint like that we assign to indexedDB directly, for
  // different reasons.
  /* $FlowExpectError */ /* eslint-disable-next-line no-global-assign */
  indexedDB = new FDBFactory();
}
beforeEach(resetIndexedDb);
afterEach(resetIndexedDb);

const listOfProfileInformations = [
  {
    profileToken: '0123456789',
    jwtToken: null,
    publishedDate: new Date('4 Jul 2020 14:00'), // "today" earlier
    name: '',
    preset: null,
    originHostname: null,
    meta: {
      product: 'Fennec',
      // Not more meta information, to test that we can handle profiles that
      // don't have this information.
    },
    urlPath: '/',
    publishedRange: { start: 1000, end: 3000 },
  },
  {
    profileToken: 'ABCDEFGHI',
    jwtToken: null,
    publishedDate: new Date('3 Jul 2020 08:00'), // yesterday
    name: 'Layout profile',
    preset: 'web',
    originHostname: null,
    meta: {
      product: 'Firefox',
      platform: 'X11',
      toolkit: 'gtk',
      oscpu: 'Linux x86_64',
      misc: 'rv:68.0',
    },
    urlPath: '/',
    publishedRange: { start: 1000, end: 1005 },
  },
  {
    profileToken: '123abc456',
    jwtToken: null,
    publishedDate: new Date('20 May 2018'), // ancient date
    name: '',
    preset: null,
    originHostname: 'https://www.cnn.com',
    meta: {
      product: 'Firefox Preview',
      platform: 'Android 7.0',
      toolkit: 'android',
      misc: 'rv:80.0',
      oscpu: 'Linux armv7l',
    },
    urlPath:
      '/public/123abc456/calltree/?globalTrackOrder=0-1-2-3-4-5-6-7-8-9-10&hiddenGlobalTracks=1-2-3-4-5-6-7-8-9&localTrackOrderByPid=32027-1-2-0~12629-0~12149-0~11755-0~11945-0~12533-0~11798-0~11862-0~12083-0~12118-0~12298-0-1~&search=bla&thread=11&v=5',
    // This tests ranges that should show up as Âµs
    publishedRange: { start: 40000, end: 40000.1 },
  },
  {
    profileToken: 'WINDOWS',
    jwtToken: null,
    publishedDate: new Date('4 Jul 2020 13:00'),
    name: 'Another good profile',
    preset: null,
    originHostname: 'https://profiler.firefox.com',
    meta: {
      product: 'Firefox',
      platform: 'Windows',
      toolkit: 'windows',
      misc: 'rv:77.0',
      oscpu: 'Windows NT 10.0; Win64; x64',
    },
    urlPath: '/public/WINDOWS/marker-chart/',
    publishedRange: { start: 2000, end: 40000 },
  },
  {
    profileToken: 'MACOSX',
    jwtToken: null,
    publishedDate: new Date('5 Jul 2020 11:00'), // This is the future!
    name: 'MacOS X profile',
    preset: null,
    originHostname: 'https://mozilla.org',
    meta: {
      product: 'Firefox',
      platform: 'Macintosh',
      toolkit: 'cocoa',
      misc: 'rv:62.0',
      oscpu: 'Intel Mac OS X 10.12',
    },
    urlPath: '/public/MACOSX/marker-chart/',
    publishedRange: { start: 2000, end: 40000 },
  },
];

async function storeProfileInformations(listOfProfileInformations) {
  for (const profileInfo of listOfProfileInformations) {
    await storeProfileData(profileInfo);
  }
}

describe('ListOfPublishedProfiles', () => {
  function setup(props) {
    // Because the rendering is dependent on the timezone, all dates in this
    // test are using a timezone-dependent format so that the rendering is the
    // same in all timezones.
    mockDate('4 Jul 2020 15:00'); // Now is 4th of July, at 3pm local timezone.

    const store = blankStore();
    const renderResult = render(
      <Provider store={store}>
        <ListOfPublishedProfiles withActionButtons={false} {...props} />
      </Provider>
    );

    const { getByText } = renderResult;

    function getAllRecordingsLink() {
      return getByText(/all your recordings/);
    }

    return {
      ...renderResult,
      getAllRecordingsLink,
    };
  }

  it('limits the number of rendered profiles with the limit prop', async () => {
    // 1. Add some profiles to the local indexeddb.
    await storeProfileInformations(listOfProfileInformations);

    // 2. Render the component and test.
    const { container, findByText, getAllRecordingsLink } = setup({ limit: 3 });
    await findByText(/macOS/);
    expect(container.querySelectorAll('li')).toHaveLength(3);
    getAllRecordingsLink(); // This shouldn't throw.
    expect(container.firstChild).toMatchSnapshot();
  });

  it('does not display the link to all recordings when there is 3 profiles or less', async () => {
    await storeProfileInformations(listOfProfileInformations.slice(0, 3));
    const { findByText, getAllRecordingsLink } = setup({ limit: 3 });
    await findByText(/Layout profile/);
    expect(() => getAllRecordingsLink()).toThrow('Unable to find an element');
  });
});
