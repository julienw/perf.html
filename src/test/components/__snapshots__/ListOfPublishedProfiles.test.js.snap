// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListOfPublishedProfiles limits the number of rendered profiles with the limit prop 1`] = `
<ul
  class="publishedProfilesList"
>
  <li
    class="publishedProfilesListItem"
  >
    <a
      class="publishedProfilesLink"
      href="http://localhost//public/MACOSX/marker-chart/"
      title="Click here to load profile MacOS X profile"
    >
      <div
        class="publishedProfilesDate"
      >
        Jul 5, 2020
      </div>
      <div
        class="publishedProfilesInfo"
      >
        <div
          class="publishedProfilesName"
        >
          <strong>
            MacOS X profile
          </strong>
           (
          38.0s
          )
        </div>
        <div
          class="profileMetaInfoSummary"
        >
          <div
            class="profileMetaInfoSummaryProductAndVersion"
            data-product="Firefox"
          >
            Firefox 62
          </div>
          <div
            class="profileMetaInfoSummaryPlatform"
            data-toolkit="macOS 10.12"
          >
            macOS 10.12
          </div>
        </div>
      </div>
    </a>
  </li>
  <li
    class="publishedProfilesListItem"
  >
    <a
      class="publishedProfilesLink"
      href="http://localhost//"
      title="Click here to load profile #012345"
    >
      <div
        class="publishedProfilesDate"
      >
        2:00 PM
      </div>
      <div
        class="publishedProfilesInfo"
      >
        <div
          class="publishedProfilesName"
        >
          <strong>
            Profile #012345
          </strong>
           (
          2.0s
          )
        </div>
        <div
          class="profileMetaInfoSummary"
        >
          <div
            class="profileMetaInfoSummaryProductAndVersion"
            data-product="Fennec"
          >
            Fennec
          </div>
          <div
            class="profileMetaInfoSummaryPlatform"
            data-toolkit=""
          />
        </div>
      </div>
    </a>
  </li>
  <li
    class="publishedProfilesListItem"
  >
    <a
      class="publishedProfilesLink"
      href="http://localhost//public/WINDOWS/marker-chart/"
      title="Click here to load profile Another good profile"
    >
      <div
        class="publishedProfilesDate"
      >
        1:00 PM
      </div>
      <div
        class="publishedProfilesInfo"
      >
        <div
          class="publishedProfilesName"
        >
          <strong>
            Another good profile
          </strong>
           (
          38.0s
          )
        </div>
        <div
          class="profileMetaInfoSummary"
        >
          <div
            class="profileMetaInfoSummaryProductAndVersion"
            data-product="Firefox"
          >
            Firefox 77
          </div>
          <div
            class="profileMetaInfoSummaryPlatform"
            data-toolkit="Windows 10"
          >
            Windows 10
          </div>
        </div>
      </div>
    </a>
  </li>
</ul>
`;
